### ISSUE TITLE
[SQLsmith] ERROR:  operator is not unique: "char" = character

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#17871]
### What version of Materialize are you using?

v0.45.0-dev (f10e2c3)

### How did you install Materialize?

Built from source

### What is the issue?

```
materialize=> SELECT NULLIF(subq_0.c3, CAST(NULL AS char(1))) AS c3 FROM (SELECT ref_0.confdeltype AS c3 FROM pg_catalog.pg_constraint AS ref_0) AS subq_0;
ERROR:  operator is not unique: "char" = character
HINT:  Could not choose a best candidate operator. You might need to add explicit type casts.
```
Previously this was considered just a bad error message, this one also seems a bit unclear to me: https://github.com/MaterializeInc/materialize/issues/1808
So this is probably nothing serious. I'll just ignore it in SQLsmith since it's only a query error.

ci-regexp: operator is not unique: "char"
ci-regexp: operator is not unique: character

**Comment by def-** [comment]
Easier repro:
```
materialize=> select 'a'::character = 'a'::"char";
ERROR:  operator is not unique: character = "char"
HINT:  Could not choose a best candidate operator. You might need to add explicit type casts.
```

**Comment by def-** [comment]
pg catalog objects seem to use `"char"` type, while the rest uses `character`. Could we use the same character type for all?

**Comment by sploiselle** [comment]
This is caused by our function selection treating all string-like types as strings when determining which function to use ([code](https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/plan/typeconv.rs#L970))--unfortunately, not as easy as just ripping this out.

**Comment by def-** [comment]
Then I guess we need to implement the relevant functions for all combinations? I'm not sure how much effort this is.

In Postgres:
```
deen=> select 'a'::character = 'a'::"char";
 ?column?
----------
 t
(1 row)
```

**Comment by sploiselle** [comment]
We'll need to fix the type system to more adequately understand the relationship between the different types in the string category. The fix is probably a 0.5-1 day's worth of work; I tried just yanking out the bit that is causing the issue but that results in an error somewhere else in planning that wasn't immediately obvious to me--some expression somewhere resolves to the wrong type.
